### PR TITLE
TMDM-14236 Wrong result returned by REST API PUT /data/{containerName}/query

### PR DESCRIPTION
**Pull request by zfkang-talend**
JIRA: https://jira.talendforge.org/browse/TMDM-14236

**What is the current behavior?** (You should also link to an open issue here)
Contains the repeated data records for full-text search.

**What is the new behavior?**
1. generate the correct index value for all field
2. modify the logic for get the correct query path.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/LuceneQueryGenerator.java [hunk 1]
```diff
@@ -79,13 +79,16 @@
...
 import java.util.StringTokenizer;
+import java.util.stream.Collectors;
 
 class LuceneQueryGenerator extends VisitorAdapter<Query> {
 
     private static final String FUZZY_SEARCH = "lucene.fuzzy.search"; //$NON-NLS-1$
 
     private final Collection<ComplexTypeMetadata> types;
 
+    private final Collection<ComplexTypeMetadata> selectTypes;
```
Comment (hw-zhu) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/LuceneQueryGenerator.java]: selectedTypes

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/LuceneQueryGenerator.java [hunk 2]
```diff
@@ -450,7 +454,23 @@ private String getFullPathName(FieldFullText fieldFullText) {
...
         }
-        return fullPathFieldName.toString();
+        return fullPathFieldName;
+    }
+
+    private void getFullPathName(ComplexTypeMetadata complexTypeMetadata, FieldMetadata fieldMetadata, StringBuilder sb) {
+        complexTypeMetadata.getFields().stream().filter(field -> field instanceof ReferenceFieldMetadata)
+                .collect(Collectors.toList()).forEach(field -> {
+            if (((ReferenceFieldMetadata) field).getReferencedType().equals(fieldMetadata.getContainingType())) {
+                sb.append(getFullPathName(field));
```
Comment (hw-zhu) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/LuceneQueryGenerator.java]: I see it's possible to throw FullTextQueryCompositeKeyException for above method `getFullPathName()`, how does we deal with it here?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/LuceneQueryGenerator.java [hunk 3]
```diff
@@ -466,9 +486,28 @@ public Query visit(FieldFullText fieldFullText) {
...
         }
         // TMDM-13918 Query by field of joining type, field name in lucene index should be fkIdField.fieldName
-        else if (fieldMetadata instanceof SimpleTypeFieldMetadata && !types.contains(fieldMetadata.getContainingType())) {
+        else if (fieldMetadata instanceof SimpleTypeFieldMetadata && !types.contains(fieldMetadata.getContainingType())
+                && !selectTypes.contains(fieldMetadata.getContainingType())) {
             FieldMetadata keyField = fieldMetadata.getContainingType().getKeyFields().iterator().next();
             fieldName = keyField.getName() + "." + fieldName; //$NON-NLS-1$
+        } else if ((!selectTypes.contains(fieldMetadata.getContainingType()) && types.contains(fieldMetadata.getContainingType()))
+                || (!types.contains(fieldMetadata.getContainingType()) && selectTypes
+                .contains(fieldMetadata.getContainingType()))) {
```
Comment (hw-zhu) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/LuceneQueryGenerator.java]: Can you refine the condition? it's a little complex to understand.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/ReferenceEntityBridge.java [hunk 4]
```diff
@@ -282,7 +282,15 @@ public void handle(String name, Object value, Field field, Document document, Lu
                 LOGGER.debug("insert a new index record with key-value pair [ " + name + "." + field.getName() + ":"
                         + value.toString());
             }
-            luceneOptions.addFieldToDocument(name + "." + field.getName(), MultilingualIndexHandler.getIndexedContent(value.toString()), document);
+
+            if (value.toString().startsWith("[") && value.toString().endsWith("]")) {
+                luceneOptions.addFieldToDocument(name + "." + field.getName(), MultilingualIndexHandler.getIndexedContent(value.toString()), document);
+            } else {
+                String[] values = value.toString().split(" ");
```
Comment (hw-zhu) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/ReferenceEntityBridge.java]: //$NON-NLS-1$

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/LuceneQueryGenerator.java [hunk 5]
```diff
@@ -450,26 +454,79 @@ private String getFullPathName(FieldFullText fieldFullText) {
...
+                .collect(Collectors.toList()).forEach(field -> {
+            if (((ReferenceFieldMetadata) field).getReferencedType().equals(fieldMetadata.getContainingType())) {
+                ReferenceFieldMetadata referenceFieldMetadata = ((ReferenceFieldMetadata) fieldMetadata);
+                if (referenceFieldMetadata.getReferencedType().getKeyFields().size() > 1) {
+                    throw new FullTextQueryCompositeKeyException(referenceFieldMetadata.getReferencedType().getName());
+                } else {
+                    sb.append(getFullPathName(field));
+                    sb.append(fieldMetadata.getName());
+                }
+
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/LuceneQueryGenerator.java]: Remove this useless blank line.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/ReferenceEntityBridge.java [hunk 6]
```diff
@@ -282,7 +282,15 @@ public void handle(String name, Object value, Field field, Document document, Lu
...
+                for(String splitValue : values){
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/ReferenceEntityBridge.java]: Format the code

File: org.talend.mdm.core.storage.sql/src/test/java/com/amalto/core/query/StorageFullTextTest.java [hunk 7]
```diff
@@ -2233,6 +2250,113 @@ public void testFuzzySearchCountry() throws Exception {
         MDMConfiguration.getConfiguration().setProperty(luceneFuzzySearch, "false");
     }
 
+    public void testOneToManySearch() {
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/test/java/com/amalto/core/query/StorageFullTextTest.java]: Format the code

**Comment by zfkang-talend** [comment]
Selenium test: 
- passed
- detail: http://192.168.31.54:9999/job/MASTER_MDM_WEB_Windows_Firefox_DEBUG/22/

REST Test:
- passed
- detail: http://192.168.31.210:8080/view/Master/job/MDM-REST-EE-MySQL8-CentOS-DEBUG2/418/

SOAP Test:
- passed,
- detail: http://192.168.31.210:8080/view/Master/job/MDM-SOAP-EE-CentOS-MySQL8-DEBUG1/398/

**Pull request merged**